README.md
# ReactNative BIP39 (react-native-bip39)

[![Build Status](https://img.shields.io/github/workflow/status/mfellner/react-native-bip39/test?style=flat&colorA=000000&colorB=000000)](https://github.com/mfellner/react-native-bip39/actions?query=workflow%3Atest)
[![Codecov](https://img.shields.io/codecov/c/github/mfellner/react-native-bip39?colorA=000000&colorB=000000)](https://app.codecov.io/gh/mfellner/react-native-bip39)
[![Version](https://img.shields.io/npm/v/@mfellner/react-native-bip39?style=flat&colorA=000000&colorB=000000)](https://www.npmjs.com/package/@mfellner/react-native-bip39)
[![Downloads](https://img.shields.io/npm/dt/@mfellner/react-native-bip39.svg?style=flat&colorA=000000&colorB=000000)](https://www.npmjs.com/package/@mfellner/react-native-bip39)

ReactNative-ready fork of the [original Javascript implementation of Bitcoin BIP39](https://github.com/bitcoinjs/bip39): Mnemonic code for generating deterministic keys

## Installation

```sh
yarn add \
  @mfellner/react-native-bip39 \
  @mfellner/react-native-fast-create-hash \
  react-native-fast-crypto \
  react-native-securerandom
```

## Features

- Native generation of random bytes using [react-native-randombytes](https://github.com/mvayngrib/react-native-randombytes)
- Partial implementation of node's `crypto` for react-native, using [react-native-crypto](https://github.com/mvayngrib/react-native-crypto)

## Reminder for developers

**_Please remember to allow recovery from mnemonic phrases that have invalid checksums (or that you don't have the wordlist)_**

When a checksum is invalid, warn the user that the phrase is not something generated by your app, and ask if they would like to use it anyway. This way, your app only needs to hold the wordlists for your supported languages, but you can recover phrases made by other apps in other languages.

However, there should be other checks in place, such as checking to make sure the user is inputting 12 words or more separated by a space. ie. `phrase.trim().split(/\s+/g).length >= 12`

## Examples

```js

import bip39 from '@mfellner/react-native-bip39'

// Generate a random mnemonic (uses react-native-randombytes under the hood), defaults to 128-bits of entropy

// NOTE: react-native-bip39's generateMnemonic in async, while original bip39's is sync
// this is due to react-native-randombytes which is forced to return natively generated
// bytes with a callback
static generateMnemonic = async () => {
  try {
    return await bip39.generateMnemonic(256) // default to 128
  } catch(e) {
    return false
  }
}
// => 'reveal man culture nominee tag abuse keen behave refuse warfare crisp thunder valve knock unique try fold energy torch news thought access hawk table'

bip39.mnemonicToSeedHex('basket actual')
// => '5cf2d4a8b0355e90295bdfc565a022a409af063d5365bb57bf74d9528f494bfa4400f53d8349b80fdae44082d7f9541e1dba2b003bcfec9d0d53781ca676651f'

bip39.mnemonicToSeed('basket actual')
// => <Buffer 5c f2 d4 a8 b0 35 5e 90 29 5b df c5 65 a0 22 a4 09 af 06 3d 53 65 bb 57 bf 74 d9 52 8f 49 4b fa 44 00 f5 3d 83 49 b8 0f da e4 40 82 d7 f9 54 1e 1d ba 2b ...>

bip39.validateMnemonic(myMnemonic)
// => true

bip39.validateMnemonic('basket actual')
// => false
```

```js
import bip39 from '@mfellner/react-native-bip39';

// defaults to BIP39 English word list
// uses HEX strings for entropy
const mnemonic = bip39.entropyToMnemonic('133755ff');
// => basket rival lemon

// reversible
bip39.mnemonicToEntropy(mnemonic);
// => '133755ff'
```


CHANGELOG.md


## [2.3.0-beta.2](https://github.com/mfellner/react-native-bip39/compare/v2.3.0-beta.1...v2.3.0-beta.2) (2022-04-19)


### Bug Fixes

* include src in dist dir ([e9c3286](https://github.com/mfellner/react-native-bip39/commit/e9c328693e7ad215fae4fc46f5e1b40ea39f81da))## [2.3.0-beta.1](https://github.com/mfellner/react-native-bip39/compare/v2.3.0-beta.0...v2.3.0-beta.1) (2022-04-17)


### Bug Fixes

* use native hash function ([d4628d7](https://github.com/mfellner/react-native-bip39/commit/d4628d74b321f230c94dc49c7389669b0fe96e8e))## [2.3.0-beta.0](https://github.com/mfellner/react-native-bip39/compare/v2.2.0...v2.3.0-beta.0) (2022-04-10)

## [2.2.0](https://github.com/mfellner/react-native-bip39/compare/v2.2.0...v2.3.0-beta.0) (2015-08-17)

### [2.1.2](https://github.com/mfellner/react-native-bip39/compare/v2.2.0...v2.3.0-beta.0) (2015-04-08)

### [2.1.1](https://github.com/mfellner/react-native-bip39/compare/v2.2.0...v2.3.0-beta.0) (2014-09-28)


### Reverts

* Revert "package: specify reliance on latest browserify" ([84dca8c](https://github.com/mfellner/react-native-bip39/commit/84dca8ccf6c5fa0640e97df139a44ecea410a5de))

## [2.1.0](https://github.com/mfellner/react-native-bip39/compare/v2.2.0...v2.3.0-beta.0) (2014-09-24)

## [2.0.0](https://github.com/mfellner/react-native-bip39/compare/v2.2.0...v2.3.0-beta.0) (2014-08-17)

## [1.2.0](https://github.com/mfellner/react-native-bip39/compare/v2.2.0...v2.3.0-beta.0) (2014-07-06)

## [1.1.0](https://github.com/mfellner/react-native-bip39/compare/v2.2.0...v2.3.0-beta.0) (2014-05-28)

### [1.0.2](https://github.com/mfellner/react-native-bip39/compare/v2.2.0...v2.3.0-beta.0) (2014-03-15)

### [1.0.1](https://github.com/mfellner/react-native-bip39/compare/v2.2.0...v2.3.0-beta.0) (2014-03-14)

## [0.1.0](https://github.com/mfellner/react-native-bip39/compare/v2.2.0...v2.3.0-beta.0) (2014-03-10)

test/index.test.ts
import { Buffer } from 'buffer';
import {
  createHash as mockCreateHash,
  pbkdf2 as mockPbkdf2,
  randomBytes as mockRandomBytes,
} from 'crypto';
import * as bip39 from '../src/';
import DEFAULT_WORDLIST from '../wordlists/en.json';
import vectors from './vectors.json';
import CUSTOM_WORDLIST from './wordlist.json';

const mockGenerateSecureRandomAsBase64 = jest.fn(async (length: number) =>
  mockRandomBytes(length).toString('base64'),
);

const mockBuffer = Buffer;

jest.mock('react-native', () => ({
  Platform: {
    select: () => '',
  },
  NativeModules: {
    RNSecureRandom: {
      generateSecureRandomAsBase64: (length: number) => mockGenerateSecureRandomAsBase64(length),
    },
    RNFastCrypto: {
      pbkdf2Sha512: jest.fn((data: string, salt: string, iterations: number, size: number) => {
        return new Promise((resolve, reject) => {
          mockPbkdf2(
            mockBuffer.from(data, 'base64'),
            mockBuffer.from(salt, 'base64'),
            iterations,
            size,
            'sha512',
            (err, derivedKey) => {
              if (err) return reject(err);
              resolve(derivedKey.toString('base64'));
            },
          );
        });
      }),
    },
    FastCreateHash: {
      createHash: jest.fn(async (data64: string, algorithm: string): Promise<string> => {
        const inp = global.Buffer.from(data64, 'base64');
        return mockCreateHash(algorithm).update(inp).digest('base64');
      }),
    },
  },
}));

test.each([
  ...vectors.english.map((v, i) => ['English', undefined, 'TREZOR', v, i] as const),
  ...vectors.japanese.map(
    (v, i) => ['Japanese', bip39.wordlists.JA, '㍍ガバヴァぱばぐゞちぢ十人十色', v, i] as const,
  ),
  ...vectors.custom.map((v, i) => ['Custom', CUSTOM_WORDLIST, undefined, v, i] as const),
])('for %s test vector %#', async (_description, wordlist, password, v) => {
  const ventropy = v[0];
  const vmnemonic = v[1];
  const vseedHex = v[2];

  expect(await bip39.mnemonicToEntropy(vmnemonic, wordlist)).toEqual(ventropy);
  // TODO FIXME
  expect(await bip39.mnemonicToSeedHex(vmnemonic, password)).toEqual(vseedHex);

  expect(await bip39.entropyToMnemonic(ventropy, wordlist)).toEqual(vmnemonic);

  const rng = async () => Buffer.from(ventropy, 'hex');

  expect(await bip39.generateMnemonic(undefined, rng, wordlist)).toEqual(vmnemonic);
  expect(await bip39.validateMnemonic(vmnemonic, wordlist)).toBe(true);
});

test.each(vectors.japanese)('UTF8 passwords', async (_ventropy, vmnemonic, vseedHex) => {
  const password = '㍍ガバヴァぱばぐゞちぢ十人十色';
  const normalizedPassword = 'メートルガバヴァぱばぐゞちぢ十人十色';

  expect(await bip39.mnemonicToSeedHex(vmnemonic, password)).toEqual(vseedHex); // 'mnemonicToSeedHex normalizes passwords',

  expect(await bip39.mnemonicToSeedHex(vmnemonic, normalizedPassword)).toEqual(vseedHex); // 'mnemonicToSeedHex leaves normalizes passwords as-is',
});

test('README example 1', async () => {
  const entropy = 'ffffffffffffffffffffffffffffffff';
  const mnemonic = await bip39.entropyToMnemonic(entropy);

  expect(mnemonic).toEqual('zoo zoo zoo zoo zoo zoo zoo zoo zoo zoo zoo wrong');

  // reversible
  expect(await bip39.mnemonicToEntropy(mnemonic)).toEqual(entropy);
});

test('README example 2', async () => {
  mockGenerateSecureRandomAsBase64.mockImplementationOnce(async (length) => {
    return Buffer.from('qwertyuiopasdfghjklzxcvbnm[];,./'.slice(0, length)).toString('base64');
  });

  const mnemonic = await bip39.generateMnemonic();

  expect(mnemonic).toEqual(
    'imitate robot frame trophy nuclear regret saddle around inflict case oil spice',
  );
  expect(await bip39.validateMnemonic(mnemonic)).toBe(true);
});

test('README example 3', async () => {
  const mnemonic = 'basket actual';
  const seed = await bip39.mnemonicToSeed(mnemonic);
  const seedHex = await bip39.mnemonicToSeedHex(mnemonic);

  expect(seed.toString('hex')).toEqual(seedHex);

  expect(seedHex).toEqual(
    '5cf2d4a8b0355e90295bdfc565a022a409af063d5365bb57bf74d9528f494bfa4400f53d8349b80fdae44082d7f9541e1dba2b003bcfec9d0d53781ca676651f',
  );
  expect(await bip39.validateMnemonic(mnemonic)).toBe(false);
});

test('invalid entropy', async () => {
  await expect(bip39.entropyToMnemonic(Buffer.from('', 'hex'))).rejects.toThrow(
    /^Invalid entropy$/,
  ); // 'throws for empty entropy')

  await expect(bip39.entropyToMnemonic(Buffer.from('000000', 'hex'))).rejects.toThrow(
    /^Invalid entropy$/,
  ); // "throws for entropy that's not a multitude of 4 bytes",

  await expect(
    bip39.entropyToMnemonic(Buffer.from(new Array(1028 + 1).join('00'), 'hex')),
  ).rejects.toThrow(/^Invalid entropy$/); // 'throws for entropy that is larger than 1024',
});

test('generateMnemonic can vary entropy length', async () => {
  const words = (await bip39.generateMnemonic(160)).split(' ');

  expect(words.length).toBe(15); // 'can vary generated entropy bit length');
});

test('generateMnemonic only requests the exact amount of data from an RNG', async () => {
  expect.assertions(1);

  await bip39.generateMnemonic(160, async (size) => {
    expect(size).toEqual(160 / 8);
    return Buffer.alloc(size);
  });
});

test('generateMnemonic rejects invalid entropy', async () => {
  expect(bip39.generateMnemonic(6)).rejects.toThrowError(/^Invalid entropy$/);
});

test('validateMnemonic', async () => {
  await expect(bip39.validateMnemonic('sleep kitten')).resolves.toBe(false); // 'fails for a mnemonic that is too short');
  await expect(bip39.validateMnemonic('sleep kitten sleep kitten sleep kitten')).resolves.toBe(
    false,
  ); // 'fails for a mnemonic that is too short',
  await expect(
    bip39.validateMnemonic(
      'turtle front uncle idea crush write shrug there lottery flower risky shell',
    ),
  ).resolves.toBe(false); // 'fails if mnemonic words are not in the word list',
  await expect(
    bip39.validateMnemonic(
      'sleep kitten sleep kitten sleep kitten sleep kitten sleep kitten sleep kitten',
    ),
  ).resolves.toBe(false); // 'fails for invalid checksum',
});

test('exposes standard wordlists', () => {
  expect(bip39.wordlists.EN.length).toEqual(2048);
  expect(typeof bip39.wordlists.EN[0]).toBe('string');
  expect(bip39.wordlists.EN).toEqual(DEFAULT_WORDLIST);
});


example/metro.config.js
const path = require('path');
const blockList = require('metro-config/src/defaults/exclusionList');
const escape = require('escape-string-regexp');
const pak = require('../package.json');

const root = path.resolve(__dirname, '..');

const modules = Object.keys({
  ...pak.peerDependencies,
});

module.exports = {
  projectRoot: __dirname,
  watchFolders: [root],

  // We need to make sure that only one version is loaded for peerDependencies
  // So we blockList them at the root, and alias them to the versions in example's node_modules
  resolver: {
    blockList: blockList(
      modules.map(
        (m) =>
          new RegExp(`^${escape(path.join(root, 'node_modules', m))}\\/.*$`)
      )
    ),

    extraNodeModules: modules.reduce((acc, name) => {
      acc[name] = path.join(__dirname, 'node_modules', name);
      return acc;
    }, {}),
  },

  transformer: {
    getTransformOptions: async () => ({
      transform: {
        experimentalImportSupport: false,
        inlineRequires: true,
      },
    }),
  },
};


example/android/app/src/main/java/com/example/MainApplication.java
package com.example;

import android.app.Application;
import android.content.Context;
import com.facebook.react.PackageList;
import com.facebook.react.ReactApplication;
import com.facebook.react.ReactInstanceManager;
import com.facebook.react.ReactNativeHost;
import com.facebook.react.ReactPackage;
import com.facebook.react.config.ReactFeatureFlags;
import com.facebook.soloader.SoLoader;
import com.example.newarchitecture.MainApplicationReactNativeHost;
import java.lang.reflect.InvocationTargetException;
import java.util.List;
// import com.reactnativebip39.ReactNativeBip39Package;

public class MainApplication extends Application implements ReactApplication {

  private final ReactNativeHost mReactNativeHost =
      new ReactNativeHost(this) {
        @Override
        public boolean getUseDeveloperSupport() {
          return BuildConfig.DEBUG;
        }

        @Override
        protected List<ReactPackage> getPackages() {
          @SuppressWarnings("UnnecessaryLocalVariable")
          List<ReactPackage> packages = new PackageList(this).getPackages();
          // Packages that cannot be autolinked yet can be added manually here.
          // packages.add(new ReactNativeBip39Package());
          return packages;
        }

        @Override
        protected String getJSMainModuleName() {
          return "index";
        }
      };

  private final ReactNativeHost mNewArchitectureNativeHost =
      new MainApplicationReactNativeHost(this);

  @Override
  public ReactNativeHost getReactNativeHost() {
    if (BuildConfig.IS_NEW_ARCHITECTURE_ENABLED) {
      return mNewArchitectureNativeHost;
    } else {
      return mReactNativeHost;
    }
  }

  @Override
  public void onCreate() {
    super.onCreate();
    // If you opted-in for the New Architecture, we enable the TurboModule system
    ReactFeatureFlags.useTurboModules = BuildConfig.IS_NEW_ARCHITECTURE_ENABLED;
    SoLoader.init(this, /* native exopackage */ false);
    initializeFlipper(this, getReactNativeHost().getReactInstanceManager());
  }

  /**
   * Loads Flipper in React Native templates. Call this in the onCreate method with something like
   * initializeFlipper(this, getReactNativeHost().getReactInstanceManager());
   *
   * @param context
   * @param reactInstanceManager
   */
  private static void initializeFlipper(
      Context context, ReactInstanceManager reactInstanceManager) {
    if (BuildConfig.DEBUG) {
      try {
        /*
         We use reflection here to pick up the class that initializes Flipper,
        since Flipper library is not available in release mode
        */
        Class<?> aClass = Class.forName("com.example.ReactNativeFlipper");
        aClass
            .getMethod("initializeFlipper", Context.class, ReactInstanceManager.class)
            .invoke(null, context, reactInstanceManager);
      } catch (ClassNotFoundException e) {
        e.printStackTrace();
      } catch (NoSuchMethodException e) {
        e.printStackTrace();
      } catch (IllegalAccessException e) {
        e.printStackTrace();
      } catch (InvocationTargetException e) {
        e.printStackTrace();
      }
    }
  }
}


src/index.ts
import { createHash } from '@mfellner/react-native-fast-create-hash';
import { Buffer } from 'buffer';
import { pbkdf2 } from 'react-native-fast-crypto';
import { generateSecureRandom } from 'react-native-securerandom';
import unorm from 'unorm';
import DEFAULT_WORDLIST from '../wordlists/en.json';
import SPANISH_WORDLIST from '../wordlists/es.json';
import JAPANESE_WORDLIST from '../wordlists/ja.json';

declare global {
  interface Uint8Array {
    toString(encoding?: string): string;
  }
}

type RandomNumberGenerator = (size: number) => Promise<Uint8Array>;

const INVALID_MNEMONIC = 'Invalid mnemonic';
const INVALID_ENTROPY = 'Invalid entropy';
const INVALID_CHECKSUM = 'Invalid mnemonic checksum';

export async function mnemonicToSeed(mnemonic: string, password?: string) {
  const mnemonicBuffer = Buffer.from(normalize(mnemonic), 'utf8');
  const saltBuffer = Buffer.from(salt(normalize(password)), 'utf8');
  return pbkdf2.deriveAsync(mnemonicBuffer, saltBuffer, 2048, 64, 'sha512');
}

export async function mnemonicToSeedHex(mnemonic: string, password = '') {
  const seed = await mnemonicToSeed(mnemonic, password);
  return seed.toString('hex');
}

export async function mnemonicToEntropy(mnemonic: string, wordlist: string[] = DEFAULT_WORDLIST) {
  const words = normalize(mnemonic).split(' ');
  if (words.length % 3 !== 0) {
    throw new Error(INVALID_MNEMONIC);
  }
  // convert word indices to 11 bit binary strings
  const bits = words
    .map((word) => {
      const index = wordlist.indexOf(word);
      if (index === -1) {
        throw new Error(INVALID_MNEMONIC);
      }
      return lpad(index.toString(2), '0', 11);
    })
    .join('');
  // split the binary string into ENT/CS
  const dividerIndex = Math.floor(bits.length / 33) * 32;
  const entropyBits = bits.slice(0, dividerIndex);
  const checksumBits = bits.slice(dividerIndex);
  // calculate the checksum and compare
  const entropyBytes = entropyBits.match(/(.{1,8})/g)?.map(binaryToByte) || [];
  if (entropyBytes.length < 16) {
    throw new Error(INVALID_ENTROPY);
  }
  if (entropyBytes.length > 32) {
    throw new Error(INVALID_ENTROPY);
  }
  if (entropyBytes.length % 4 !== 0) {
    throw new Error(INVALID_ENTROPY);
  }
  const entropy = Buffer.from(entropyBytes);
  const newChecksum = await deriveChecksumBits(entropy);
  if (newChecksum !== checksumBits) {
    throw new Error(INVALID_CHECKSUM);
  }
  return entropy.toString('hex');
}

export async function entropyToMnemonic(entropy: string | Buffer, wordlist: string[] = DEFAULT_WORDLIST) {
  if (typeof entropy === 'string') {
    entropy = Buffer.from(entropy, 'hex');
  }
  // 128 <= ENT <= 256
  if (entropy.length < 16) {
    throw new TypeError(INVALID_ENTROPY);
  }
  if (entropy.length > 32) {
    throw new TypeError(INVALID_ENTROPY);
  }
  if (entropy.length % 4 !== 0) {
    throw new TypeError(INVALID_ENTROPY);
  }
  const entropyBits = bytesToBinary(Array.from(entropy));
  const checksumBits = await deriveChecksumBits(entropy);
  const bits = entropyBits + checksumBits;
  const chunks = bits.match(/(.{1,11})/g);
  if (!chunks) {
    throw new TypeError(INVALID_ENTROPY);
  }
  const words = chunks.map((binary) => {
    const index = binaryToByte(binary);
    return wordlist[index];
  });
  return wordlist[0] === '\u3042\u3044\u3053\u304f\u3057\u3093' // Japanese wordlist
    ? words.join('\u3000')
    : words.join(' ');
}

export async function generateMnemonic(
  strength = 128,
  rng: RandomNumberGenerator = generateSecureRandom,
  wordlist: string[] = DEFAULT_WORDLIST,
) {
  if (strength % 32 !== 0) {
    throw new TypeError(INVALID_ENTROPY);
  }

  const bytes = await rng(strength / 8);
  const hexBuffer = Buffer.from(bytes);

  return entropyToMnemonic(hexBuffer, wordlist);
}

export async function validateMnemonic(mnemonic: string, wordlist?: string[]) {
  try {
    await mnemonicToEntropy(mnemonic, wordlist);
  } catch (e) {
    return false;
  }
  return true;
}

function salt(password: string) {
  return 'mnemonic' + (unorm.nfkd(password) || ''); // Use unorm until String.prototype.normalize gets better browser support
}

//=========== helper methods from bitcoinjs-lib ========

function bytesToBinary(bytes: number[]): string {
  return bytes.map((x: number): string => lpad(x.toString(2), '0', 8)).join('');
}

function lpad(str: string, padString: string, length: number): string {
  while (str.length < length) {
    str = padString + str;
  }
  return str;
}

function normalize(str = ''): string {
  return str.normalize('NFKD');
}

function binaryToByte(bin: string): number {
  return parseInt(bin, 2);
}

async function deriveChecksumBits(entropyBuffer: Buffer) {
  const ENT = entropyBuffer.length * 8;
  const CS = ENT / 32;
  const hash = await createHash(entropyBuffer, 'sha256');
  return bytesToBinary(Array.from(hash)).slice(0, CS);
}

export const wordlists = {
  EN: DEFAULT_WORDLIST,
  ES: SPANISH_WORDLIST,
  JA: JAPANESE_WORDLIST,
};
